File: problem_solving_with_DSA/linkedList/C/getNthNode.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Definition for a linked list node.
 */
typedef struct node
{
   int data;
   struct node *next;
}lNode ;


/**
 * getNthNode - Get the Nth Node data in a linked list
 * 
 * @head: Head pointer
 * @index: Index of the node to get
 * Return: Data of the node at the index
 */
int getNthNode(lNode *head, int index)
{
    lNode *temp = head;
    
    if (head != NULL)
    {
        while (temp && index)
        {
            temp = temp->next;
            index--;
        }
        if (temp == NULL)
            return (-1);
        return temp->data;
    }
    return (-1);
}

/**
 * insertAtEnd - Insert a new node to the back of a linked list
 * 
 * @head: Head pointer 
 * @value: Value to insert
 * Return: Pointer to the newNode
 */
lNode *insertAtEnd(lNode **head, int value)
{
    lNode *newNode;
    
    newNode = malloc(sizeof(lNode));
    
    if (newNode)
    {
        newNode->data = value;
        newNode->next = NULL;
        
        if (*(head) == NULL)
        {
            *(head) = newNode;
            return newNode;
        }
        else
        {
            lNode *last = *(head);
            while (last->next)
            {
                last = last->next;
            }
            last->next = newNode;        
        }
    }
    return newNode;
    
}

/**
 * printNode - Print Node data
 * 
 * @head: Head pointer 
 */
void printNode(lNode *head)
{
    lNode *temp = head;
    
    while(temp)
    {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}


int main(void)
{
    lNode *head = NULL;
    
    insertAtEnd(&head, 10);
    insertAtEnd(&head, 20);
    insertAtEnd(&head, 30);
    insertAtEnd(&head, 40);
    insertAtEnd(&head, 50);
    
    printNode(head);
    
    printf("%d\n",getNthNode(head, 9));
    
    return (0);
}

File: problem_solving_with_DSA/linkedList/python/getNthNode.py
"""
Get a middle node in a linked list
"""

class Node:
    """Representation of a linked list Node"""
    def __init__(self, data) -> None:
        self.data = data
        self.next = None
    
class LinkedList:
    """Representation of a linked list"""
    def __init__(self) -> None:
        self.head = None
    
    def createNode(self, value):
        """Create a new node 
        
        Args:
            value(int) - Data to insert
        """
        newNode = Node(value)
        
        if self.head == None:
            self.head = newNode
            return newNode
        else:
            last = self.head
            while last.next:
                last = last.next
            last.next = newNode
        return newNode
        
    def printLinkedList(self):
        """Print Data in a linkedlist"""
        temp = self.head
        
        while temp != None:
            print("{} ".format(temp.data), end="")
            temp = temp.next
        print()
    
    def getNthNode(self, index):
        """ Get Nth node in a linked list
        
        Args:
            index(int): Index of the node
            
        Return (int): Data at the index
        """
        if self.head != None:
            temp = self.head
            while temp and index:
               temp = temp.next
               index -= 1
            if temp == None:
                return (-1)
            return temp.data       
        return (-1)
        


if __name__ == "__main__":
    llist = LinkedList()
    llist.createNode(10)
    llist.createNode(20)
    llist.createNode(30)
    llist.createNode(40)
    llist.createNode(50)
    
    llist.printLinkedList()
    print(llist.getNthNode(1))
